File: week12/7/[241122](Leet225)_Implemnet Stack using Queues.py
"""Implement a last-in-first-out (LIFO) stack using only two queues. The implemented stack should support all the functions of a normal stack (push, top, pop, and empty).

Implement the MyStack class:

void push(int x) Pushes element x to the top of the stack.
int pop() Removes the element on the top of the stack and returns it.
int top() Returns the element on the top of the stack.
boolean empty() Returns true if the stack is empty, false otherwise.
Notes:

You must use only standard operations of a queue, which means that only push to back, peek/pop from front, size and is empty operations are valid.
Depending on your language, the queue may not be supported natively. You may simulate a queue using a list or deque (double-ended queue) as long as you use only a queue's standard operations."""

"""아이디어
    Stack안에 store_Queue, temp_queue을 만든다.
    push할 때 store_queue에 있는 요소들을 모든 요소를 temp_queue으로 옮기고 temp_queue를 store_queue로 바꿔주면 됨!
    근데 시간을 아끼려면 temp_queue이 비어있을 때만 store-queue에서 옮기는 작업을 하면 됨.
"""
"""Develeoped Amswer(하나의 queue만 사용) : 
    deqeue에 넣고, 맨 앞의 요소( 먼저 들어와 있는 것)를 제외하고 나머지는 뺐다가 다시 넣음."""

## 두 개의 qeuue 사용
from collections import deque

class MyStack:
    def __init__(self):
        self.store_q = deque()
        self.temp_q = deque()
        
    def push(self, x: int) -> None:
        self.store_q.append(x)
        
    def pop(self) -> int:
        n = len(self.store_q)
        for _ in range(n-1):
            x = self.store_q.popleft()
            self.temp_q.append(x)
        
        popped_element = self.store_q.popleft()
        
        #swap q1 and q2    
        self.store_q, self.temp_q = self.temp_q, self.store_q
        
        return popped_element

    def top(self) -> int:
        n = len(self.store_q)
        for _ in range(n-1):
            x = self.store_q.popleft()
            self.temp_q.append(x)
        
        top_element = self.store_q[0]
        self.store_q.popleft()
        
        self.temp_q.append(top_element)
        
        #swap q1 and q2    
        self.store_q, self.temp_q = self.temp_q, self.store_q
        
        return top_element
        
    def empty(self) -> bool: # store_q에 요소가 있으면 false, 요소가 없으면 true
        return False if self.store_q else True






##---- 하나의 queue만 사용
from collections import deque

class MyStack:
    def __init__(self):
        self.store_q = deque()
        
    def push(self, x: int) -> None:
        self.store_q.append(x)
        n = len(self.store_q)
        for _ in range(n-1):
            x = self.store_q.popleft()
            self.store_q.append(x)

    def pop(self) -> int:  
        return self.store_q.popleft()

    def top(self) -> int:        
        return self.store_q[0]
        
    def empty(self) -> bool: # store_q에 요소가 있으면 false, 요소가 없으면 true
        return False if self.store_q else True
        
        


# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()

File: week12/7/[241122](Leet225)_Implemnet Stack using Queues.c
/*
Constraints:
    1 <= x <= 9
    At most 100 calls will be made to push, pop, top, and empty.
    All the calls to pop and top are valid.
 */

//Using 2 Queues

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAX_SIZE 100

typedef struct {
    int* data;
    int front;
    int rear;
    int size;
    int capacity;
} Queue;

// 큐 생성
Queue* createQueue(int capacity) {
    Queue* queue = (Queue*)malloc(sizeof(Queue));
    queue->data = (int*)malloc(capacity * sizeof(int));
    queue->front = 0;
    queue->rear = 0;
    queue->size = 0;
    queue->capacity = capacity;
    return queue;
}

// 큐 연산
bool isEmpty(Queue* queue) {
    return queue->size == 0;
}

void enqueue(Queue* queue, int value) {
    if (queue->size == queue->capacity) return; // 용량 초과 방지
    queue->data[queue->rear] = value;
    queue->rear = (queue->rear + 1) % queue->capacity;
    queue->size++;
}

int dequeue(Queue* queue) {
    if (isEmpty(queue)) return -1; // 비어있을 때
    int value = queue->data[queue->front];
    queue->front = (queue->front + 1) % queue->capacity;
    queue->size--;
    return value;
}

int front(Queue* queue) {
    return queue->data[queue->front];
}

// 큐 메모리 해제
void freeQueue(Queue* queue) {
    free(queue->data);
    free(queue);
}



typedef struct {
    Queue* primaryQueue;
    Queue* secondaryQueue;
} MyStack;


MyStack* myStackCreate() {
    MyStack* stack = (MyStack*)malloc(sizeof(MyStack));
    stack->primaryQueue = createQueue(MAX_SIZE);
    stack->secondaryQueue = createQueue(MAX_SIZE);
    return stack;
}

void myStackPush(MyStack* obj, int x) {
    enqueue(obj->secondaryQueue, x);

    // primaryQueue의 모든 요소를 secondaryQueue로 이동
    while (!isEmpty(obj->primaryQueue)) {
        enqueue(obj->secondaryQueue, dequeue(obj->primaryQueue));
    }

    // primaryQueue와 secondaryQueue를 교환
    Queue* temp = obj->primaryQueue;
    obj->primaryQueue = obj->secondaryQueue;
    obj->secondaryQueue = temp;
}

int myStackPop(MyStack* obj) {
    return dequeue(obj->primaryQueue);
}

int myStackTop(MyStack* obj) {
    return front(obj->primaryQueue);
}

bool myStackEmpty(MyStack* obj) {
    return isEmpty(obj->primaryQueue);
}

void myStackFree(MyStack* obj) {
    freeQueue(obj->primaryQueue);
    freeQueue(obj->secondaryQueue);
    free(obj);
}

/**
 * Your MyStack struct will be instantiated and called as such:
 * MyStack* obj = myStackCreate();
 * myStackPush(obj, x);
 * int param_2 = myStackPop(obj);
 * int param_3 = myStackTop(obj);
 * bool param_4 = myStackEmpty(obj);
 * myStackFree(obj);
*/


// Using 1 queue---------------------------------------------------------------------------------------------------
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct {
    int* data;
    int front;
    int rear;
    int size;
    int capacity;
} Queue;

// 큐 생성
Queue* createQueue(int capacity) {
    Queue* queue = (Queue*)malloc(sizeof(Queue));
    queue->data = (int*)malloc(capacity * sizeof(int));
    queue->front = 0;
    queue->rear = 0;
    queue->size = 0;
    queue->capacity = capacity;
    return queue;
}

// 큐 연산
bool isEmpty(Queue* queue) {
    return queue->size == 0;
}

void enqueue(Queue* queue, int value) {
    if (queue->size == queue->capacity) return; // 용량 초과 방지
    queue->data[queue->rear] = value;
    queue->rear = (queue->rear + 1) % queue->capacity;
    queue->size++;
}

int dequeue(Queue* queue) {
    if (isEmpty(queue)) return -1; // 비어있을 때
    int value = queue->data[queue->front];
    queue->front = (queue->front + 1) % queue->capacity;
    queue->size--;
    return value;
}

int front(Queue* queue) {
    return queue->data[queue->front];
}

// 큐 메모리 해제
void freeQueue(Queue* queue) {
    free(queue->data);
    free(queue);
}

// MyStack 구조체 정의
typedef struct {
    Queue* queue;
} MyStack;

// MyStack 생성
MyStack* myStackCreate() {
    MyStack* stack = (MyStack*)malloc(sizeof(MyStack));
    stack->queue = createQueue(100); // 초기 용량 설정
    return stack;
}

// 스택 연산
void myStackPush(MyStack* obj, int x) {
    enqueue(obj->queue, x);
    int size = obj->queue->size;

    // 새로 삽입된 요소를 앞부분으로 이동
    for (int i = 0; i < size - 1; i++) {
        enqueue(obj->queue, dequeue(obj->queue));
    }
}

int myStackPop(MyStack* obj) {
    return dequeue(obj->queue);
}

int myStackTop(MyStack* obj) {
    return front(obj->queue);
}

bool myStackEmpty(MyStack* obj) {
    return isEmpty(obj->queue);
}

void myStackFree(MyStack* obj) {
    freeQueue(obj->queue);
    free(obj);
}